R/data.table.R

```diff
@@ -292,119 +301,99 @@ replace_dot_alias = function(e) {
-      # Replacing the NA symbol makes DT[NA] and DT[c(1,NA)] consistent and provides
-      # an easy way to achieve a single row of NA as users expect rather than requiring them
-      # to know and change to DT[NA_integer_].
-      isub=NA_integer_
-    }
     isnull_inames = FALSE
     # Fixes 4994: a case where quoted expression with a "!", ex: expr = quote(!dt1); dt[eval(expr)] requires
     # the "eval" to be checked before `as.name("!")`. Therefore interchanged.
-    restore.N = remove.N = FALSE
-    if (exists(".N", envir=parent.frame(), inherits=FALSE)) {
-      old.N = get(".N", envir=parent.frame(), inherits=FALSE)
-      locked.N = bindingIsLocked(".N", parent.frame())
-      if (locked.N) eval(call("unlockBinding", ".N", parent.frame()))  # eval call to pass R CMD check NOTE until we find cleaner way
-      assign(".N", nrow(x), envir=parent.frame(), inherits=FALSE)
-      restore.N = TRUE
-      # the comment below is invalid hereafter (due to fix for #1145)
-      # binding locked when .SD[.N] but that's ok as that's the .N we want anyway
-
-      # TO DO: change isub at C level s/.N/nrow(x); changing a symbol to a constant should be ok
+    if (!is.language(isub)) {
+        if (identical(isub, NA)) {
+        # if (is.logical(isub) && length(isub) == 1L && is.na(isub) && !is.matrix(isub)) {
+        # only possibility *isub* can be NA (logical) is the symbol NA itself; i.e. DT[NA]
+        # replace NA in this case with NA_integer_ as that's almost surely what user intended to
+        # return a single row with NA in all columns. (DT[0] returns an empty table, with correct types.)
+        # Any expression (including length 1 vectors) that evaluates to a single NA logical will
+        # however be left as NA logical since that's important for consistency to return empty in that
+        # case; e.g. DT[Col==3] where DT is 1 row and Col contains NA.
+        # Replacing the NA symbol makes DT[NA] and DT[c(1,NA)] consistent and provides
+        # an easy way to achieve a single row of NA as users expect rather than requiring them
+        # to know and change to DT[NA_integer_].
+        i = isub=NA_integer_
+        } else i = isub
     } else {
-       assign(".N", nrow(x), envir=parent.frame(), inherits=FALSE)
-       remove.N = TRUE
-    }
-    if (isub %iscall% "eval") {  # TO DO: or ..()
-      isub = eval(.massagei(isub[[2L]]), parent.frame(), parent.frame())
-      if (is.expression(isub)) isub=isub[[1L]]
-    }
-    if (isub %iscall% "!") {
-      notjoin = TRUE
-      if (!missingnomatch) stop("not-join '!' prefix is present on i but nomatch is provided. Please remove nomatch.");
-      nomatch = 0L
-      isub = isub[[2L]]
-      # #932 related so that !(v1 == 1) becomes v1 == 1 instead of (v1 == 1) after removing "!"
-      if (isub %iscall% "(" && !is.name(isub[[2L]]))
+      if (isub %iscall% "eval") {  # TO DO: or ..()
+        isub = eval(.massagei(isub[[2L]]), list(.N = nrow(x)), parent.frame())
```

Comment:
can we add `.SD=x` to `envir` arg here & get `.SD` to work in `i` just like that?

Reply:
I just compiled with adding ```.SD``` and success! 

Note, previously ```.N``` was assigned to the ```parent.frame()``` and then restoring it if necessary. Because of that, all 4 ```eval``` calls related to processing ```i``` were largely the same.

While skipping that approach is faster, we now have to deal with associating each of the 4 ```eval``` calls with ```.N``` or whatever special variable(s) we want to use so there's a little more accounting. In theory we could have also used the previous approach to also assign ```.SD``` to the parent.frame.

Reply:
ha, came across this comment again :upside_down_face: